Add converter for Arctica-M N1 RSR data.

simonrp84 opened the pull request:
This adds spectral response functions for the MSU-GS/A instruments aboard the Arctica-M N1 satellite.
There are two instruments aboard the satellite, which have differing spectral response functions. Here I add them as `det-1` and `det-2`.

To achieve this, I have also updated `utils.py` to add the ability to save HDF5 files by detector - which wasn't implemented before.

 - [x] Tests passed: Passes ``pytest pyspectral``
 - [x] Passes ``flake8 pyspectral`` 
 - [x] Fully documented

pnuu commented on pyspectral/utils.py:
```diff
@@ -244,7 +244,7 @@ def sort_data(x_vals, y_vals):
 
 
-def convert2hdf5(ClassIn, platform_name, bandnames, scale=1e-06):
+def convert2hdf5(ClassIn, platform_name, bandnames, scale=1e-06, dets=None):
```
I think the new kwarg should be `detectors` for clarity. And have a short explanation in the docstring.

pnuu commented on rsr_convert_scripts/msu_gsa_reader.py:
```diff
@@ -0,0 +1,106 @@
+def main():
+    """Main"""
+    for platform_name in ['Arctica-M-N1', ]:
+        tohdf5(MsugsaRSR, platform_name, MSUGSA_BAND_NAMES, dets=['det-1', 'det-2'])
```
And also remember to update the kwarg naming here.

adybbroe commented:
@simonrp84 You can update `utils.py` with this url and bump the version to 1.0.18:
`https://zenodo.org/record/6026563/files/pyspectral_rsr_data.tgz`

simonrp84 commented:
Thanks for the comments @pnuu and @adybbroe! I think I've updated / addressed everything now. I also added `Arctica-M-N1` to the `INSTRUMENTS` dict in `utils.py`.

adybbroe commented on pyspectral/utils.py:
```diff
@@ -67,13 +67,14 @@
                }
 
-HTTP_PYSPECTRAL_RSR = "https://zenodo.org/record/5797957/files/pyspectral_rsr_data.tgz"
+HTTP_PYSPECTRAL_RSR = "https://zenodo.org/record/6026563/files/pyspectral_rsr_data.tg"
```
Should be `tgz` as file extension!

adybbroe reviewed (changes requested):
Sorry @simonrp84 But could you please add some testing on your added code in `utils.py`? The coverage went down. I didn't pay attention to that previously....

simonrp84 commented:
> Sorry @simonrp84 But could you please add some testing on your added code in `utils.py`? The coverage went down. I didn't pay attention to that previously....
Looking at what's changed, it seems that the function I altered didn't have any tests before - so I would need to add them. This will probably not be a quick effort as I've not worked with mocking HDF5 before. Will see what I can do.

adybbroe commented on pyspectral/utils.py:
```diff
@@ -264,17 +266,38 @@ def convert2hdf5(ClassIn, platform_name, bandnames, scale=1e-06):
-            rsp = sensor.rsr['response'][~np.isnan(sensor.rsr['wavelength'])]
-            grp.attrs['central_wavelength'] = get_central_wave(wvl, rsp)
-            arr = sensor.rsr['wavelength']
+
+            # If multiple detectors, assume all have same wavelength range in SRF.
+            if detectors is not None:
+                wvl = sensor.rsr[detectors[0]]['wavelength'][~np.isnan(sensor.rsr[detectors[0]]['wavelength'])]
+                arr = sensor.rsr[detectors[0]]['wavelength']
+                grp.attrs['number_of_detectors'] = len(detectors)
```
This seems untested. Could we improve that?

adybbroe commented on pyspectral/utils.py:
```diff
@@ -264,17 +266,38 @@ def convert2hdf5(ClassIn, platform_name, bandnames, scale=1e-06):
+                wvl = sensor.rsr[detectors[0]]['wavelength'][~np.isnan(sensor.rsr[detectors[0]]['wavelength'])]
+                arr = sensor.rsr[detectors[0]]['wavelength']
+                grp.attrs['number_of_detectors'] = len(detectors)
+            else:
+                wvl = sensor.rsr['wavelength'][~np.isnan(sensor.rsr['wavelength'])]
+                arr = sensor.rsr['wavelength']
```
Also here, this is not tested it seems.

adybbroe commented on pyspectral/utils.py:
```diff
@@ -264,17 +266,38 @@ def convert2hdf5(ClassIn, platform_name, bandnames, scale=1e-06):
-            arr = sensor.rsr['response']
-            dset = grp.create_dataset('response', arr.shape, dtype='f')
-            dset[...] = arr
+
+            # Now to do the responses
+            if detectors is None:
+                rsp = sensor.rsr['response'][~np.isnan(sensor.rsr['wavelength'])]
+                grp.attrs['central_wavelength'] = get_central_wave(wvl, rsp)
+                arr = sensor.rsr['response']
+                dset = grp.create_dataset('response', arr.shape, dtype='f')
+                dset[...] = arr
```
Same here, is this tested?

adybbroe commented on pyspectral/utils.py:
```diff
@@ -264,17 +266,38 @@ def convert2hdf5(ClassIn, platform_name, bandnames, scale=1e-06):
-            arr = sensor.rsr['response']
-            dset = grp.create_dataset('response', arr.shape, dtype='f')
-            dset[...] = arr
+
+            # Now to do the responses
+            if detectors is None:
+                rsp = sensor.rsr['response'][~np.isnan(sensor.rsr['wavelength'])]
+                grp.attrs['central_wavelength'] = get_central_wave(wvl, rsp)
+                arr = sensor.rsr['response']
+                dset = grp.create_dataset('response', arr.shape, dtype='f')
+                dset[...] = arr
+            else:
+                for cur_det in detectors:
+                    det_grp = grp.create_group(cur_det)
+                    rsp = sensor.rsr[cur_det]['response'][~np.isnan(sensor.rsr[cur_det]['wavelength'])]
+                    det_grp.attrs['central_wavelength'] = get_central_wave(wvl, rsp)
+                    arr = sensor.rsr[cur_det]['response']
+                    dset = det_grp.create_dataset('response', arr.shape, dtype='f')
+                    dset[...] = arr
```
Again, not covered by tests I think.

adybbroe reviewed (commented):
There are changes to utils.py which seems poorly covered by tests. Is it something you can try improve? Or do you need help?

adybbroe commented:
@simonrp84 I will try see if I can help our here. Maybe as Arctica-M is kind of on hold as we have no real-time flow of data this PR is not urgent and we can see if we find some time improving the test coverage of the `convert2hdf5` function before merging. Ok?

simonrp84 commented:
To be honest I've forgotten what I did in this PR! Will check through it next week and then we can figure out how to move on it. Like you say, no rush now.

simonrp84 commented:
I have now added some tests that should improve the coverage of the code previously missed. I chose to not mock the `h5py` library but to actually reader an HDF5 RSR file through the tests and then delete it after. Am unsure if this is the best way to do things so please let me know what you think!

adybbroe reviewed (approved):
LGTM
